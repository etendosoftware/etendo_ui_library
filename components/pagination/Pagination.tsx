/* Imports */
import React, {useState} from 'react';

import {Text, View, TouchableOpacity} from 'react-native';
import {PaginationStyleVariant} from './Pagination.styles';
import {PaginationProps} from './Pagination.types';
import {
  arrowFirst,
  arrowLast,
  arrowPickerHorizontalIcon,
} from '../../assets/images/icons';
import CustomImage from '../../SecondaryComponents/CustomImage';

/* Pagination component */
export const Pagination = ({
  currentPage,
  onChangeSelected,
  totalData,
  amountDataPerPage,
  pagination,
}: PaginationProps) => {
  // Will serve to set the page number where the user is
  const [page, setPage] = useState<number>(currentPage);

  // This constant will count the number of pages that the data table will have based on the number of row
  // that the progammer wishes to display per page
  const pageNumbers: Array<number> = [];
  for (let i = 1; i <= Math.ceil(totalData / amountDataPerPage); i++) {
    pageNumbers.push(i);
  }

  /* Helper functions to page up or down */
  const backToFirstPage = () => {
    setPage(1);
  };

  const backPage = (page: number) => {
    setPage(page - 1);
  };

  const nextPage = (page: number) => {
    setPage(page + 1);
  };

  const nextLastPage = (pageNumbers: Array<number>) => {
    setPage(pageNumbers.length);
  };

  return (
    <View style={PaginationStyleVariant.primary.container}>
      <View style={PaginationStyleVariant.primary.container}>
        <TouchableOpacity
          disabled={page <= 1 || totalData === 0}
          onPress={() => {
            /* Allows to detect the selected item */
            backToFirstPage();
            onChangeSelected(1, 0);
          }}
        >
          <CustomImage
            src={arrowFirst}
            alt="back-double-button"
            source={{uri: arrowFirst}}
            style={{
              width: 10,
              height: 10,
              marginRight: 10,
            }}
          />
        </TouchableOpacity>

        <TouchableOpacity
          disabled={page <= 1 || totalData === 0}
          onPress={() => {
            backPage(page);
            if (page !== 1) {
              onChangeSelected(page - 1, pagination - amountDataPerPage);
            }
          }}
        >
          <CustomImage
            src={arrowPickerHorizontalIcon}
            alt="back-button"
            source={{uri: arrowPickerHorizontalIcon}}
            style={{
              width: 10,
              height: 10,
              marginRight: 10,
              rotate: '180deg',
            }}
          />
        </TouchableOpacity>

        {/* Text with layout to display the page number the user is on */}
        <View style={PaginationStyleVariant.primary.pageNumberContainer}>
          <Text style={PaginationStyleVariant.primary.pageNumberText}>
            {page}
          </Text>
        </View>

        <TouchableOpacity
          disabled={page === pageNumbers.length || totalData === 0}
          onPress={() => {
            nextPage(page);
            if (page !== Math.ceil(totalData / 8)) {
              onChangeSelected(page + 1, pagination + amountDataPerPage);
            }
          }}
        >
          <CustomImage
            style={{
              width: 10,
              height: 10,
              marginRight: 10,
            }}
            src={arrowPickerHorizontalIcon}
            alt="next-button"
            source={{uri: arrowPickerHorizontalIcon}}
          />
        </TouchableOpacity>

        <TouchableOpacity
          disabled={page === pageNumbers.length || totalData === 0}
          onPress={() => {
            nextLastPage(pageNumbers);
            if (page !== Math.ceil(totalData / 8)) {
              onChangeSelected(
                Math.ceil(totalData / amountDataPerPage),
                totalData - (totalData % amountDataPerPage) - 8,
              );
            }
          }}
        >
          <CustomImage
            style={{
              width: 10,
              height: 10,
              marginRight: 10,
            }}
            src={arrowLast}
            alt="next-double-button"
            source={{uri: arrowLast}}
          />
        </TouchableOpacity>
      </View>
    </View>
  );
};
