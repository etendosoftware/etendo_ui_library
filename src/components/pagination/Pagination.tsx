/* Imports */
import React, { useState } from 'react';

import { View, TouchableOpacity, Text } from 'react-native';
import { PaginationStyleVariant } from './Pagination.styles';
import { PaginationProps } from './Pagination.types';
import { ArrowLeftIcon } from '../../assets/images/icons/ArrowLeftIcon';
import { ArrowFirstIcon } from '../../assets/images/icons/ArrowFirstIcon';
import { ArrowRightIcon } from '../../assets/images/icons/ArrowRightIcon';
import { ArrowLastIcon } from '../../assets/images/icons/ArrowLastIcon';

/* Pagination component */
const Pagination = ({
  currentPage,
  onChangeSelected,
  totalData,
  amountDataPerPage,
  pagination,
}: PaginationProps) => {
  // Will serve to set the page number where the user is
  const [page, setPage] = useState<number>(currentPage);

  // This constant will count the number of pages that the data table will have based on the number of row
  // that the progammer wishes to display per page
  const pageNumbers: Array<number> = [];
  for (let i = 1; i <= Math.ceil(totalData / amountDataPerPage); i++) {
    pageNumbers.push(i);
  }

  /* Helper functions to page up or down */
  const backToFirstPage = () => {
    setPage(1);
  };

  const backPage = (page: number) => {
    setPage(page - 1);
  };

  const nextPage = (page: number) => {
    setPage(page + 1);
  };

  const nextLastPage = (pageNumbers: Array<number>) => {
    setPage(pageNumbers.length);
  };
  return (
    <View style={PaginationStyleVariant.primary.container}>
      <View style={PaginationStyleVariant.primary.container}>
        <TouchableOpacity
          disabled={page <= 1 || totalData === 0}
          onPress={() => {
            /* Allows to detect the selected item */
            backToFirstPage();
            onChangeSelected(1, 0);
          }}>
          <ArrowFirstIcon
            style={{
              width: 10,
              height: 10,
              marginRight: 10,
            }}
          />
        </TouchableOpacity>

        <TouchableOpacity
          disabled={page <= 1 || totalData === 0}
          onPress={() => {
            backPage(page);
            if (page !== 1) {
              onChangeSelected(page - 1, pagination - amountDataPerPage);
            }
          }}>
          <ArrowLeftIcon
            style={{
              height: 10,
              width: 10,
              marginRight: 10,
            }}
          />
        </TouchableOpacity>

        {/* Text with layout to display the page number the user is on */}
        <View style={PaginationStyleVariant.primary.pageNumberContainer}>
          <Text style={PaginationStyleVariant.primary.pageNumberText}>
            {page}
          </Text>
        </View>

        <TouchableOpacity
          disabled={page === pageNumbers.length || totalData === 0}
          onPress={() => {
            nextPage(page);
            if (page !== Math.ceil(totalData / 8)) {
              onChangeSelected(page + 1, pagination + amountDataPerPage);
            }
          }}>
          <ArrowRightIcon
            style={{
              height: 10,
              width: 10,
              marginRight: 10,
            }}
          />
        </TouchableOpacity>

        <TouchableOpacity
          disabled={page === pageNumbers.length || totalData === 0}
          onPress={() => {
            nextLastPage(pageNumbers);
            if (page !== Math.ceil(totalData / 8)) {
              onChangeSelected(
                Math.ceil(totalData / amountDataPerPage),
                totalData - (totalData % amountDataPerPage) - 8,
              );
            }
          }}>
          <ArrowLastIcon
            style={{
              height: 10,
              width: 10,
              marginRight: 10,
            }}
          />
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default Pagination;
