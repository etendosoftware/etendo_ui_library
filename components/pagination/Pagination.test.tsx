import React from 'react';
import renderer from 'react-test-renderer';

import {Pagination} from './Pagination';

describe('Running Test for Tab', () => {
  it('Check Tab Disabled', () => {
    const tree = renderer
      .create(
        <Pagination currentPage={1} totalData={40} amountDataPerPage={8} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
