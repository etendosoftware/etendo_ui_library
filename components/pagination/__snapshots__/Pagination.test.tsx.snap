// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running Test for Tab Check Tab Disabled 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "height": 22,
      "width": 120,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "height": 22,
        "width": 120,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/images/back-double-button.png",
          }
        }
        style={
          Object {
            "height": 22,
            "marginRight": 17,
            "width": 20,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/images/back-button.png",
          }
        }
        style={
          Object {
            "height": 22,
            "marginRight": 17,
            "width": 20,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#dddff6",
          "borderRadius": 5,
          "height": 27,
          "justifyContent": "center",
          "marginRight": 17,
          "width": 27,
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "fontWeight": "bold",
          }
        }
      >
        1
      </Text>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/images/next-button.png",
          }
        }
        style={
          Object {
            "height": 22,
            "marginRight": 17,
            "width": 20,
          }
        }
      />
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/images/next-double-button.png",
          }
        }
        style={
          Object {
            "height": 22,
            "marginRight": 17,
            "width": 20,
          }
        }
      />
    </View>
  </View>
</View>
`;
